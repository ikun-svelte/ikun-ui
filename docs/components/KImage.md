---
title: KImage
lang: en-US
---

# KImage

Besides the native features of img, support lazy load, custom placeholder and load failure, etc.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/image
pnpm add @ikun-ui/image-view
```

```bash [yarn]
yarn add @ikun-ui/image
yarn add @ikun-ui/image-view
```

```bash [npm]
npm install @ikun-ui/image
npm install @ikun-ui/image-view
```

:::

## Basic usage

Indicate how the image should be resized to fit its container by fit, same as native [object-fit <span class="i-carbon-link text-12px" />](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit).

<demo src="image/basic.svelte"  github='Image'></demo>

## Placeholder

You can use the `placeholder` slot to define the placeholder content when the image is not loaded.

<demo src="image/placeholder.svelte"  github='Image'></demo>

## Load Failed

You can use the `error` slot to define the placeholder content when the image fails to load.

<demo src="image/failed.svelte"  github='Image'></demo>

## Lazy Load

The lazy loading function can be turned on through `lazy`,
and the image will not be loaded until it is scrolled into the visible range.

<demo src="image/lazy.svelte"  github='Image'></demo>

## Image Preview

The function of previewing large images can be enabled through `previewSrcList`

<demo src="image/preview.svelte"  github='Image'></demo>

## Image Props

| Name            | Type                                                          | Default | Description                                                                                                                                                                                |
| --------------- | ------------------------------------------------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| scrollContainer | `string \| HTMLElement`                                       | `""`    | The container to add scroll listener when using lazy load. By default, the container to add scroll listener when using lazy load.                                                          |
| loading         | `'eager' \| 'lazy'`                                           | `""`    | Indicates how the browser should load the image, same as [native <span class="i-carbon-link text-12px" />](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-loading).    |
| previewSrcList  | `string[]`                                                    | `[]`    | Allow big image preview.                                                                                                                                                                   |
| fit             | `'fill' \| 'contain'    \| 'cover' \| 'none' \| 'scale-down'` | `""`    | Indicate how the image should be resized to fit its container, same as [object-fit <span class="i-carbon-link text-12px" />](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit). |
| src             | `string`                                                      | `""`    | Image source, same as native.                                                                                                                                                              |
| alt             | `string`                                                      | `""`    | Native attribute `alt`.                                                                                                                                                                    |
| cls             | `string`                                                      | `""`    | Additional class for component                                                                                                                                                             |
| attrs           | `any`                                                         | `{}`    | Additional attributes                                                                                                                                                                      |

## Image Events

| Name  | Description                       | Type                    |
| ----- | --------------------------------- | ----------------------- |
| load  | same as native load.              | `(event: Event)=> void` |
| error | same as native error..            | `(event: Event)=> void` |
| show  | trigger when the viewer displays. | `(event: Event)=> void` |

## Image Slots

| Name        | Description                                              |
| ----------- | -------------------------------------------------------- |
| placeholder | custom placeholder content when image hasn't loaded yet. |
| error       | custom image load failed content.                        |

## ImageView Props

| Name  | Type       | Default | Description                      |
| ----- | --------   | ------- | ------------------------------   |
| show  | `boolean`  | `false` | Whether to display image preview |
| urls  | `string[]` | `[]`    | Preview image link list          |
| cls   | `string`   | `""`    | Additional class for component   |
| attrs | `any`      | `{}`    | Additional attributes            |

## Image Events

| Name  | Description            | Type                    |
| ----- | ---------------------- | ----------------------- |
| close | triggered when closed. | `(event: Event)=> void` |
