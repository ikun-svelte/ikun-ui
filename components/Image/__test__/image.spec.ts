import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import KImage from '../src';
import KImageFit from './fixture/fix.svelte';
import KImagePlaceHolder from './fixture/placeholder.svelte';
import { tick } from 'svelte';
let host;

const initHost = () => {
	host = globalThis.document.createElement('div');
	host.setAttribute('id', 'host');
	globalThis.document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.useRealTimers();
});

describe('Test: KImage', () => {
	vi.mock('svelte', async () => {
		const actual = (await vi.importActual('svelte')) as object;
		return {
			...actual,
			// @ts-ignore
			onMount: (await import('svelte/internal')).onMount
		};
	});

	test('props: cls', async () => {
		//@ts-ignore
		const instance = new KImage({
			//@ts-ignore
			target: host,
			props: {
				cls: 'k-image--test'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-image--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: src', async () => {
		//@ts-ignore
		const instance = new KImage({
			target: host,
			props: {
				src: 'demo'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host!.innerHTML.includes('src="demo"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: alt', async () => {
		//@ts-ignore
		const instance = new KImage({
			target: host,
			props: {
				alt: 'demo'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host!.innerHTML.includes('alt="demo"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: fit', async () => {
		//@ts-ignore
		const instance = new KImageFit({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: previewSrcList', async () => {
		//@ts-ignore
		const instance = new KImage({
			target: host,
			props: {
				previewSrcList: ['previewSrcList']
			}
		});
		expect(instance).toBeTruthy();
		const imgBtn = host.querySelector('.k-image__inner');
		imgBtn.click();
		await tick();
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: show', async () => {
		const mockFn = vi.fn();
		//@ts-ignore
		const instance = new KImage({
			target: host,
			props: {
				previewSrcList: ['previewSrcList']
			}
		});
		expect(instance).toBeTruthy();
		//@ts-ignore
		instance.$on('show', mockFn);
		const imgBtn = host.querySelector('.k-image__inner');
		imgBtn.click();
		await tick();
		expect(mockFn).toBeCalled();
	});

	test('slots: placeHolder', async () => {
		//@ts-ignore
		const instance = new KImagePlaceHolder({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML).matchSnapshot();
	});
});
