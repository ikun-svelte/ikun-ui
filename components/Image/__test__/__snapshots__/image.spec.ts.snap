// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KImage > props: alt 1`] = `"<div class=\\"k-image\\"><img alt=\\"demo\\" aria-hidden=\\"true\\" src=\\"\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div>"`;

exports[`Test: KImage > props: cls 1`] = `"<div class=\\"k-image k-image--test\\"><img alt=\\"\\" aria-hidden=\\"true\\" src=\\"\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div>"`;

exports[`Test: KImage > props: fit 1`] = `"<div class=\\"demo-image svelte-1ep430a\\"><div class=\\"block svelte-1ep430a\\"><span class=\\"demonstration svelte-1ep430a\\">fill</span> <div class=\\"k-image\\"><img alt=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" aria-hidden=\\"true\\" style=\\"width: 100px; height: 100px; object-fit: fill;\\" src=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div> </div><div class=\\"block svelte-1ep430a\\"><span class=\\"demonstration svelte-1ep430a\\">contain</span> <div class=\\"k-image\\"><img alt=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" aria-hidden=\\"true\\" style=\\"width: 100px; height: 100px; object-fit: contain;\\" src=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div> </div><div class=\\"block svelte-1ep430a\\"><span class=\\"demonstration svelte-1ep430a\\">cover</span> <div class=\\"k-image\\"><img alt=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" aria-hidden=\\"true\\" style=\\"width: 100px; height: 100px; object-fit: cover;\\" src=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div> </div><div class=\\"block svelte-1ep430a\\"><span class=\\"demonstration svelte-1ep430a\\">none</span> <div class=\\"k-image\\"><img alt=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" aria-hidden=\\"true\\" style=\\"width: 100px; height: 100px; object-fit: none;\\" src=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div> </div><div class=\\"block svelte-1ep430a\\"><span class=\\"demonstration svelte-1ep430a\\">scale-down</span> <div class=\\"k-image\\"><img alt=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" aria-hidden=\\"true\\" style=\\"width: 100px; height: 100px; object-fit: scale-down;\\" src=\\"https://pic3.zhimg.com/80/v2-f0f6615f256bdd3aeb77ec0637bab192_1440w.webp\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div> </div></div>"`;

exports[`Test: KImage > props: previewSrcList 1`] = `
"<div class=\\"k-image\\"><img alt=\\"\\" aria-hidden=\\"true\\" src=\\"\\" class=\\"k-image__inner k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> <div class=\\"k-mask--base\\" style=\\"top: 0px; left: 0px; width: 100%; height: 100%; animation: __svelte_665066525_0 300ms linear 0ms 1 both;\\"><div class=\\"k-image-view\\"><div class=\\"k-image-view--close\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-close\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 26px; height: 26px;\\"></span></div> <div class=\\"k-image-view--body\\"><img src=\\"previewSrcList\\" alt=\\"previewSrcList\\" class=\\"k-image-view--body__img\\" style=\\"
		transform: translate3d(0px, 0px, 0px) scale3d(1, 1, 1) rotate(0deg);
		transition: transform 0.3s ease 0s;
	\\"></div> <div class=\\"k-image-view--footer__wrapper\\"><div class=\\"k-image-view--footer\\"> <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrows-vertical k-image-view--footer__icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span> <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrows-horizontal k-image-view--footer__icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span> <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-rotate-counterclockwise k-image-view--footer__icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span> <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-rotate-clockwise k-image-view--footer__icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span> <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-zoom-out k-image-view--footer__icon__disabled\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span> <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-zoom-in k-image-view--footer__icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span> </div></div></div></div></div>"
`;

exports[`Test: KImage > props: src 1`] = `"<div class=\\"k-image\\"><img alt=\\"demo\\" aria-hidden=\\"true\\" src=\\"demo\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><div class=\\"k-image__placeholder\\"></div></div> </div>"`;

exports[`Test: KImage > slots: placeHolder 1`] = `"<div class=\\"k-image\\"><img alt=\\"\\" aria-hidden=\\"true\\" style=\\"width: 100px; height: 100px;\\" src=\\"\\" class=\\"k-image__inner k-image__loading\\"> <div class=\\"k-image__wrapper\\"><p slot=\\"placeholder\\">test placeholder</p></div> </div>"`;
